Fix update password & username routes

navjeev-nagra opened the pull request:
- Eliminated username from logic, use email instead
- Has authentication with the email used during signup

Changed the PUT route to: **api/users/update/password/**

Body of request should be:

{
   email: string,
   new_password: string
   old_password: string
}

Sample:
![image](https://github.com/AnthonyOlijnyk/reservify/assets/146782694/93d83e6b-d8f8-4027-9cc2-69943a1b1b67)

I checked the DB, password seems to be updating correctly on the backend with these changes.

AnthonyOlijnyk commented on backend/UserApp/views.py:
```diff
@@ -150,36 +152,33 @@ def put(self, request, username):
 class UserUpdatePasswordView(APIView):
-    def put(self, request, username):
+    def put(self, request):
+        email = request.data.get('email', None)
```
I don't think the frontend should have to pass you an email here. If the user is logged in, then you can get the ID for that user using this function I made [here](https://github.com/AnthonyOlijnyk/reservify/blob/53bdae41178939f284b787856606451940bdb89a/backend/core/utils/authorization.py#L28).

Since the frontend people wouldn't need to pass you an email / username anymore, you should communicate with them after you make the changes that [this](https://github.com/AnthonyOlijnyk/reservify/blob/53bdae41178939f284b787856606451940bdb89a/frontend/src/pages/UserDash.js#L94-L98) should be removed, and the request body [here](https://github.com/AnthonyOlijnyk/reservify/blob/53bdae41178939f284b787856606451940bdb89a/frontend/src/pages/UserDash.js#L114-L118) should be updated to match what you are expecting in your request.

AnthonyOlijnyk commented on backend/UserApp/views.py:
```diff
@@ -150,36 +152,33 @@ def put(self, request, username):
         except User.DoesNotExist:
             return Response(
-                {'error': 'User not found'},
                 status=status.HTTP_404_NOT_FOUND
```
I know that we recently wanted to change the response format to just have a status associated with it, but Polina understandably wants some error logging to help her debug how to use the APIs. 

I would attach an error field with a message saying what happened during your request to your error responses so the frontend people have an easier time debugging what they're missing in the request.

AnthonyOlijnyk reviewed (commented):
Also, just a heads up but I was looking at the update username endpoint [here](https://github.com/AnthonyOlijnyk/reservify/blob/53bdae41178939f284b787856606451940bdb89a/backend/UserApp/views.py#L127) and it doesn't check that the user making the request to change a username has the same username inputted as the old_username field. 

This means that any user can change the username for any other user without them knowing which should probably be fixed.

navjeev-nagra commented:
### **Overall changes:**
- Changed error logging to match other views, added additional error logging cases for frontend debugging
- Validate user's existence by user ID, rather than email

### **UpdatePassword specific changes:**
- Added error logging cases with messages to

1. Ensure new and old passwords are in request body
2. Check if the current password is correct
3. Authenticate user details (email, old password)
4. Ensure new password is different from old password

PUT route for updating password: **api/users/update/password/**

### **UpdateUsername specific changes:**
- Added error logging cases with messages to

1. Ensure new and old usernames are in request body
2. Check if the user making the request matches the old_username provided (addresses issue of being able to change another user's username)
3. Ensure the new username is different from old username

- Removed username from logic to be consistent with UpdatePasswordView
- Added a check to ensure the user making the request to change a username has the same username inputted as the old_username field

UPDATED PUT route for updating username: **api/users/update/username/**

AnthonyOlijnyk commented on backend/UserApp/views.py:
```diff
@@ -124,62 +127,97 @@ def get(self, request):
-        new_username = request.data.get('new_username', None)
+                return Response({
+                'success': False,
+                'message': 'User not found.'
```
The success field isn't necessary anymore because the frontend can verify if the request succeeded by checking the status with response.ok.


Also just a small thing, but can you change the key that says `message` to be `error` instead and then don't send any message field on success? Just want some consistency for the frontend people.

AnthonyOlijnyk reviewed (approved):
Great job!!!!!
